### PR TITLE
Improve Real Time Provisioning Helper tool

### PR DESCRIPTION
**Pull request by sajjad-moradi**
## Description
Currently Real Time Provisioning Helper tool takes a completed segment as an input. With the changes in this PR, a user can provide data characteristics instead of an actual segment. With this option, the tool does a preprocessing step and generates a segment based on the provided characteristics. After the segment is generated, it just uses that segment to provide insight on the memory footprint as usual.

That main changes in the code:
- refactored a few existing `Generator`s in `data/generator` package and also added a couple of new ones
- added `Segment Generator` to `Memory Estimator`
- modified `RealtimeProvisioingHelperCommand`
## Testing Done
- Unit tests
- Ran `pinot-admin RealtimeProvisioningHelper` locally with the same files provided in the unit tests (1M rows):
```bash
2021/02/04 15:13:39.243 INFO [RealtimeProvisioningHelperCommand] [main] Executing command: RealtimeProvisioningHelper -tableConfigFile table-config.json -numPartitions 10 -pushFrequency null -numHosts 2,4,6,8,10,12,14,16 -numHours 2,3,4,5,6,7,8,9,10,11,12 -schemaWithMetadataFile schema.json -ingestionRate 150 -maxUsableHostMemory 48G -retentionHours 48
2021/02/04 15:13:41.549 INFO [MemoryEstimator$SegmentGenerator] [main] Successfully generated data file: /var/folders/sd/fgc60hhj2994pk9vm1xw235h000xqy/T/2021-02-04_15:13:39-csv/output_0.csv
2021/02/04 15:13:41.549 INFO [MemoryEstimator$SegmentGenerator] [main] Started creating segment from file: /var/folders/sd/fgc60hhj2994pk9vm1xw235h000xqy/T/2021-02-04_15:13:39-csv/output_0.csv
2021/02/04 15:13:49.084 INFO [MemoryEstimator$SegmentGenerator] [main] Successfully created segment: testTable_18667_18766_0 at directory: /var/folders/sd/fgc60hhj2994pk9vm1xw235h000xqy/T/2021-02-04_15:13:39-segment/testTable_18667_18766_0
2021/02/04 15:13:49.085 INFO [MemoryEstimator$SegmentGenerator] [main] Verifying the segment by loading it
2021/02/04 15:13:49.161 INFO [MemoryEstimator$SegmentGenerator] [main] Successfully loaded segment: testTable_18667_18766_0 of size: 18766286 bytes

============================================================
RealtimeProvisioningHelper -tableConfigFile table-config.json -numPartitions 10 -pushFrequency null -numHosts 2,4,6,8,10,12,14,16 -numHours 2,3,4,5,6,7,8,9,10,11,12 -schemaWithMetadataFile schema.json -ingestionRate 150 -maxUsableHostMemory 48G -retentionHours 48

Note:

* Table retention and push frequency ignored for determining retentionHours since it is specified in command
* See https://docs.pinot.apache.org/operators/operating-pinot/tuning/realtime
2021/02/04 15:13:53.141 INFO [RealtimeProvisioningHelperCommand] [main] 
Memory used per host (Active/Mapped)

numHosts --> 2               |4               |6               |8               |10              |12              |14              |16              |
numHours
 2 --------> 7.67G/17.86G    |4.09G/9.53G     |2.56G/5.95G     |2.05G/4.76G     |1.53G/3.57G     |1.53G/3.57G     |1.53G/3.57G     |1.02G/2.38G     |
 3 --------> 8.03G/18.22G    |4.28G/9.72G     |2.68G/6.07G     |2.14G/4.86G     |1.61G/3.64G     |1.61G/3.64G     |1.61G/3.64G     |1.07G/2.43G     |
 4 --------> 8.38G/18.58G    |4.47G/9.91G     |2.79G/6.19G     |2.24G/4.95G     |1.68G/3.72G     |1.68G/3.72G     |1.68G/3.72G     |1.12G/2.48G     |
 5 --------> 9.02G/18.93G    |4.81G/10.1G     |3.01G/6.31G     |2.41G/5.05G     |1.8G/3.79G      |1.8G/3.79G      |1.8G/3.79G      |1.2G/2.52G      |
 6 --------> 9.1G/19.29G     |4.85G/10.29G    |3.03G/6.43G     |2.43G/5.14G     |1.82G/3.86G     |1.82G/3.86G     |1.82G/3.86G     |1.21G/2.57G     |
 7 --------> 9.59G/20.5G     |5.12G/10.93G    |3.2G/6.83G      |2.56G/5.47G     |1.92G/4.1G      |1.92G/4.1G      |1.92G/4.1G      |1.28G/2.73G     |
 8 --------> 9.81G/20G       |5.23G/10.67G    |3.27G/6.67G     |2.62G/5.33G     |1.96G/4G        |1.96G/4G        |1.96G/4G        |1.31G/2.67G     |
 9 --------> 11.01G/21.21G   |5.87G/11.31G    |3.67G/7.07G     |2.94G/5.66G     |2.2G/4.24G      |2.2G/4.24G      |2.2G/4.24G      |1.47G/2.83G     |
10 --------> 10.8G/20.71G    |5.76G/11.05G    |3.6G/6.9G       |2.88G/5.52G     |2.16G/4.14G     |2.16G/4.14G     |2.16G/4.14G     |1.44G/2.76G     |
11 --------> 11.87G/21.21G   |6.33G/11.31G    |3.96G/7.07G     |3.16G/5.66G     |2.37G/4.24G     |2.37G/4.24G     |2.37G/4.24G     |1.58G/2.83G     |
12 --------> 11.23G/21.43G   |5.99G/11.43G    |3.74G/7.14G     |3G/5.71G        |2.25G/4.29G     |2.25G/4.29G     |2.25G/4.29G     |1.5G/2.86G      |
2021/02/04 15:13:53.142 INFO [RealtimeProvisioningHelperCommand] [main] 
Optimal segment size

numHosts --> 2               |4               |6               |8               |10              |12              |14              |16              |
numHours
 2 --------> 19.33M          |19.33M          |19.33M          |19.33M          |19.33M          |19.33M          |19.33M          |19.33M          |
 3 --------> 29M             |29M             |29M             |29M             |29M             |29M             |29M             |29M             |
 4 --------> 38.66M          |38.66M          |38.66M          |38.66M          |38.66M          |38.66M          |38.66M          |38.66M          |
 5 --------> 48.33M          |48.33M          |48.33M          |48.33M          |48.33M          |48.33M          |48.33M          |48.33M          |
 6 --------> 57.99M          |57.99M          |57.99M          |57.99M          |57.99M          |57.99M          |57.99M          |57.99M          |
 7 --------> 67.66M          |67.66M          |67.66M          |67.66M          |67.66M          |67.66M          |67.66M          |67.66M          |
 8 --------> 77.32M          |77.32M          |77.32M          |77.32M          |77.32M          |77.32M          |77.32M          |77.32M          |
 9 --------> 86.99M          |86.99M          |86.99M          |86.99M          |86.99M          |86.99M          |86.99M          |86.99M          |
10 --------> 96.65M          |96.65M          |96.65M          |96.65M          |96.65M          |96.65M          |96.65M          |96.65M          |
11 --------> 106.32M         |106.32M         |106.32M         |106.32M         |106.32M         |106.32M         |106.32M         |106.32M         |
12 --------> 115.98M         |115.98M         |115.98M         |115.98M         |115.98M         |115.98M         |115.98M         |115.98M         |
2021/02/04 15:13:53.144 INFO [RealtimeProvisioningHelperCommand] [main] 
Consuming memory

numHosts --> 2               |4               |6               |8               |10              |12              |14              |16              |
numHours
 2 --------> 1.16G           |632.16M         |395.1M          |316.08M         |237.06M         |237.06M         |237.06M         |158.04M         |
 3 --------> 1.66G           |904.1M          |565.06M         |452.05M         |339.04M         |339.04M         |339.04M         |226.02M         |
 4 --------> 2.15G           |1.15G           |735.02M         |588.02M         |441.01M         |441.01M         |441.01M         |294.01M         |
 5 --------> 2.65G           |1.41G           |904.98M         |723.99M         |542.99M         |542.99M         |542.99M         |361.99M         |
 6 --------> 3.15G           |1.68G           |1.05G           |859.96M         |644.97M         |644.97M         |644.97M         |429.98M         |
 7 --------> 3.65G           |1.95G           |1.22G           |995.93M         |746.94M         |746.94M         |746.94M         |497.96M         |
 8 --------> 4.15G           |2.21G           |1.38G           |1.11G           |848.92M         |848.92M         |848.92M         |565.95M         |
 9 --------> 4.64G           |2.48G           |1.55G           |1.24G           |950.9M          |950.9M          |950.9M          |633.93M         |
10 --------> 5.14G           |2.74G           |1.71G           |1.37G           |1.03G           |1.03G           |1.03G           |701.92M         |
11 --------> 5.64G           |3.01G           |1.88G           |1.5G            |1.13G           |1.13G           |1.13G           |769.9M          |
12 --------> 6.14G           |3.27G           |2.05G           |1.64G           |1.23G           |1.23G           |1.23G           |837.89M         |
2021/02/04 15:13:53.145 INFO [RealtimeProvisioningHelperCommand] [main] 
Total number of segments queried per host (for all partitions)

numHosts --> 2               |4               |6               |8               |10              |12              |14              |16              |
numHours
 2 --------> 360             |192             |120             |96              |72              |72              |72              |48              |
 3 --------> 240             |128             |80              |64              |48              |48              |48              |32              |
 4 --------> 180             |96              |60              |48              |36              |36              |36              |24              |
 5 --------> 150             |80              |50              |40              |30              |30              |30              |20              |
 6 --------> 120             |64              |40              |32              |24              |24              |24              |16              |
 7 --------> 105             |56              |35              |28              |21              |21              |21              |14              |
 8 --------> 90              |48              |30              |24              |18              |18              |18              |12              |
 9 --------> 90              |48              |30              |24              |18              |18              |18              |12              |
10 --------> 75              |40              |25              |20              |15              |15              |15              |10              |
11 --------> 75              |40              |25              |20              |15              |15              |15              |10              |
12 --------> 60              |32              |20              |16              |12              |12              |12              |8               |

```

### CODE REVIEW

File: pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java [hunk 1]
```diff
@@ -387,4 +418,182 @@ private long calculateMemoryForCompletedSegmentsPerPartition(long completedSegme
   public String[][] getNumSegmentsQueriedPerHost() {
     return _numSegmentsQueriedPerHost;
   }
+
+  private static File generateCompletedSegment(File dataCharacteristicsFile, File schemaFile, TableConfig tableConfig) {
```
Comment (siddharthteotia) [inline pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java:422]: Please make sure that dataCharacteristicsFile format is the exact same json file format that is taken by the recommendation engine (RecommenderDrivr.java). Otherwise, the follow up integration task won't work.

File: pinot-tools/src/test/resources/memory_estimation/data-characteristics.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
+{
+  "numberOfRows": 10000,
+  "columnCharacteristics": [
```
Comment (siddharthteotia) [inline pinot-tools/src/test/resources/memory_estimation/data-characteristics.json:3]: This is not same as what is taken by RecommenderDriver right? I don't think it takes columnCharacteristics.  See the file InputManager and SchemaWithMetadata.

Comment (siddharthteotia) [reply pinot-tools/src/test/resources/memory_estimation/data-characteristics.json:3]: It is fine to extend that format but hereI feel we are looking at a different format

File: pinot-tools/src/test/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimatorTest.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    File tableConfigFile = readFile("memory_estimation/table-config.json");
+    TableConfig tableConfig = JsonUtils.fileToObject(tableConfigFile, TableConfig.class);
+
+    // act
+    MemoryEstimator.SegmentGenerator
+        segmentGenerator = new MemoryEstimator.SegmentGenerator(dataCharacteristicsFile, schemaFile, tableConfig, true);
+    File generatedSegment = segmentGenerator.generate();
+
+    // assert
+    Path metadataFile = Paths.get(generatedSegment.getPath(), "v3", "metadata.properties");
```
Comment (siddharthteotia) [inline pinot-tools/src/test/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimatorTest.java:54]: how can we check for average length for string types?

File: pinot-tools/src/test/resources/memory_estimation/schema.json [hunk 4]
```diff
@@ -0,0 +1,36 @@
+{
+  "schemaName": "testTable",
+  "dimensionFieldSpecs": [
+    {
```
Comment (siddharthteotia) [inline pinot-tools/src/test/resources/memory_estimation/schema.json:4]: Let's make this as complete as possible. 

All fixed width - both single and multi-value
All variable width (string and bytes) - both single and multi value

Comment (siddharthteotia) [reply pinot-tools/src/test/resources/memory_estimation/schema.json:4]: Also, if you think about this from integration perspective, RecommenderDriver doesn't take schema. It takes the schema annotated with characteristics.  (See the files RecommenderDriver, InputManager and SchemaWithMetadata). So, something to think about for integration. If now you are providing MemoryEstimator with schema separately then I am not sure if you are going to change it back to the annotated schema in the follow-up PR when you will integrate with RecommenderDriver.

File: pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/RealtimeProvisioningHelperCommand.java [hunk 5]
```diff
@@ -71,9 +72,15 @@
...
   private String _numHours = "2,3,4,5,6,7,8,9,10,11,12";
 
-  @Option(name = "-sampleCompletedSegmentDir", required = true, metaVar = "<String>", usage = "Consume from the topic for n hours and provide the path of the segment dir after it completes")
+  @Option(name = "-sampleCompletedSegmentDir", required = false, metaVar = "<String>", usage = "Consume from the topic for n hours and provide the path of the segment dir after it completes")
   private String _sampleCompletedSegmentDir;
 
+  @Option(name = "-dataCharacteristicsFile", required = false, metaVar = "<String>", usage = "File containing characteristics by which a segment will be generated")
+  private String _dataCharacteristicsFile;
+
+  @Option(name = "-schemaFile", required = false, metaVar = "<String>")
```
Comment (siddharthteotia) [inline pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/RealtimeProvisioningHelperCommand.java:81]: We should not provide schema file. See my comment further below. The data characteristics file is essentially the schema annotated with additional info. So it alone should be enough

File: pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/DataGenerator.java [hunk 6]
```diff
@@ -163,26 +174,75 @@ private FieldSpec buildSpec(DataGeneratorSpec genSpec, String column) {
...
-      cardinality.put(col, 1000);
+    Map<String, Double> mvCountMap = new HashMap<>();
+    Map<String, Integer> lengthMap = new HashMap<>();
+    List<String> columnNames = new ArrayList<>();
+
+    int cardinalityValue = 5;
+    int strLength = 5;
+
+    String colName = "colInt";
+    dataTypes.put(colName, DataType.INT);
```
Comment (siddharthteotia) [inline pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/DataGenerator.java:193]: Sorry, I don't follow these changes. All of this information (column name, type etc) should come from the data characteristics file. What is colInt then?

File: pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/DataGeneratorSpec.java [hunk 7]
```diff
@@ -38,6 +38,8 @@
   private final Map<String, Integer> cardinalityMap;
   private final Map<String, IntRange> rangeMap;
   private final Map<String, Map<String, Object>> patternMap;
+  private final Map<String, Double> mvCountMap; // map of column name to average number of values per entry
+  private final Map<String, Integer> lengthMap; // map of column name to average length of th entry (used for string generator)
```
Comment (siddharthteotia) [inline pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/DataGeneratorSpec.java:42]: Both for string and bytes right?

File: pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/MultiValueGeneratorHelper.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+   * @param numberOfValuesPerEntry number of values per each row
+   * @param rand random object
+   * @param nextItemFunc function to get the next random item
+   * @return
+   */
+  public static List<Object> generateMultiValueEntries(double numberOfValuesPerEntry, Random rand,
+      Supplier<Object> nextItemFunc) {
+    List<Object> entries = new ArrayList<>();
+    int i = 0;
+    for (; i < numberOfValuesPerEntry - 1; i++) {
```
Comment (siddharthteotia) [inline pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/MultiValueGeneratorHelper.java:45]: (nit) I don't think we follow this coding convention. We always initialize the loop variable again in the for loop

File: pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java [hunk 9]
```diff
@@ -387,4 +418,182 @@ private long calculateMemoryForCompletedSegmentsPerPartition(long completedSegme
...
+
+      return indexDir;
+    }
+
+    private SegmentGeneratorConfig getSegmentGeneratorConfig(File csvDataFile, Schema schema, String outDir) {
+      SegmentGeneratorConfig segmentGeneratorConfig = new SegmentGeneratorConfig(_tableConfig, schema);
+      segmentGeneratorConfig.setInputFilePath(csvDataFile.getPath());
+      segmentGeneratorConfig.setFormat(FileFormat.CSV);
+      segmentGeneratorConfig.setOutDir(outDir);
+      segmentGeneratorConfig.setReaderConfig(new CSVRecordReaderConfig()); // FIXME
```
Comment (siddharthteotia) [inline pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java:553]: Why FIXME?

File: pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java [hunk 10]
```diff
@@ -107,6 +129,15 @@ public MemoryEstimator(TableConfig tableConfig, File sampleCompletedSegment, int
     _tableDataDir.mkdir();
   }
 
+  public MemoryEstimator(TableConfig tableConfig, File dataCharacteristicsFile, File schemaFile, int ingestionRate,
```
Comment (siddharthteotia) [inline pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java:132]: (nit) Add small javadoc please to highlight how this constructor is different from the other one

File: pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java (hunk 1 above)
Comment (siddharthteotia) [reply pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java:422]: separate schemaFile should not be needed

File: pinot-tools/src/test/resources/memory_estimation/data-characteristics.json (hunk 2 above)
Comment (siddharthteotia) [reply pinot-tools/src/test/resources/memory_estimation/data-characteristics.json:3]: https://github.com/apache/incubator-pinot/blob/master/pinot-controller/src/main/java/org/apache/pinot/controller/recommender/io/InputManager.java

https://github.com/apache/incubator-pinot/blob/master/pinot-controller/src/main/java/org/apache/pinot/controller/recommender/io/metadata/SchemaWithMetaData.java

File: pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java [hunk 11]
```diff
@@ -107,6 +129,15 @@ public MemoryEstimator(TableConfig tableConfig, File sampleCompletedSegment, int
...
+      long maxUsableHostMemory, int tableRetentionHours) {
+    this(tableConfig,
+        generateCompletedSegment(dataCharacteristicsFile, schemaFile, tableConfig),
```
Comment (siddharthteotia) [inline pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java:135]: Can we add a small test (or ensure manually) that the file location of the generated completed segment returned by generateCompletedSegment() is in the same segment file hierarchy as taken by the other constructor

File: pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/StringGenerator.java [hunk 12]
```diff
@@ -56,11 +65,18 @@ public void init() {
...
   @Override
   public Object next() {
+    if (numberOfValuesPerEntry == 1) {
+      return getNextString();
+    }
+    return MultiValueGeneratorHelper.generateMultiValueEntries(numberOfValuesPerEntry, rand, this::getNextString);
+  }
+
+  private String getNextString() {
     return vals.get(rand.nextInt(cardinality));
```
Comment (siddharthteotia) [inline pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/StringGenerator.java:75]: Vals is a hashmap. This is where we are going to run into issues with generation based approach where the characteristics file specifies a very high cardinality for column(s). The HashMap will likely hit OOM

File: pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/NumberGenerator.java [hunk 13]
```diff
@@ -61,23 +66,23 @@ public void init() {
...
         final long longEnd = longStart + cardinality;
         for (long i = longStart; i < longEnd; i++) {
-          longValues.add(new Long(i));
+          longValues.add(i);
         }
         break;
       case FLOAT:
         floatValues = new ArrayList<Float>();
-        final float floatStart = rand.nextFloat() * rand.nextInt(1000);
+        final float floatStart = Math.round(rand.nextFloat()* 100.0f) / 100.0f;
```
Comment (siddharthteotia) [inline pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/NumberGenerator.java:82]: why this change?

File: pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/RealtimeProvisioningHelperCommand.java (hunk 5 above)
Comment (mcvsubbu) [reply pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/RealtimeProvisioningHelperCommand.java:81]: Isnt it good to have the option of providing  a sample segment OR data characteristics?

Comment (siddharthteotia) [reply pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/RealtimeProvisioningHelperCommand.java:81]: @mcvsubbu , sorry I don't think I am following this comment? We are providing data characteristics right

File: pinot-tools/src/test/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimatorTest.java (hunk 3 above)
Comment (sajjad-moradi) [reply pinot-tools/src/test/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimatorTest.java:110]: I'm not sure! Do you have any suggestion? Although StringGenerator is a simple class and we're ok not checking that.

File: pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/DataGenerator.java (hunk 6 above)
Comment (sajjad-moradi) [reply pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/DataGenerator.java:193]: Apparently this was a test for data generator. Just wanted to add more cases to it.

File: pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/DataGeneratorSpec.java (hunk 7 above)
Comment (sajjad-moradi) [reply pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/DataGeneratorSpec.java]: Right. Updated it.

File: pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/MultiValueGeneratorHelper.java (hunk 8 above)
Comment (sajjad-moradi) [reply pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/MultiValueGeneratorHelper.java:45]: I'm using the index `i` in the next statement. I could use `numberOfValuesPerEntry - Math.floor(numberOfValuesPerEntry)` instead of `numberOfValuesPerEntry - i`, but index `i` has exactly that value and we already have it.

File: pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java (hunk 9 above)
Comment (sajjad-moradi) [reply pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java]: It was for my book-keeping that got away to get cleaned up before sending the PR. It's removed now.

File: pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java (hunk 11 above)
Comment (sajjad-moradi) [reply pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java]: Why do we want to do that? We generate it in `java.io.tmpdir` so it gets deleted automatically.

File: pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/NumberGenerator.java (hunk 13 above)
Comment (sajjad-moradi) [reply pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/NumberGenerator.java]: This makes the generated float numbers easier on the eye - with two decimal places.

File: pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/StringGenerator.java (hunk 12 above)
Comment (sajjad-moradi) [reply pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/StringGenerator.java:75]: I'm not sure why we need to generate millions of records. I assume something below 500K records should suffice and my assumption is that we won't get into trouble with that number of rows. 
BTW I have tested it on my mac with 10M rows with high cardinality on different columns in order of millions and no issue observed.

File: pinot-tools/src/test/resources/memory_estimation/schema.json (hunk 4 above)
Comment (sajjad-moradi) [reply pinot-tools/src/test/resources/memory_estimation/schema.json]: Okay, I gave up 😃  (To be fair, I wanted to do that in the next PR which is the integration part, but now it has it!)
Refactored the code to use `SchemaWithMetadata`. Now only one file is provided and that one represents both schema and data characteristics.

File: pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/StringGenerator.java (hunk 12 above)
Comment (siddharthteotia) [reply pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/StringGenerator.java:75]: It is not unusual to have million rows in a completed segment.

File: pinot-common/src/main/java/org/apache/pinot/common/data/schema/FieldMetadata.java [hunk 14]
```diff
@@ -16,24 +16,24 @@
...
 
 import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
 import com.fasterxml.jackson.annotation.JsonSetter;
 import com.fasterxml.jackson.annotation.Nulls;
 import org.apache.pinot.spi.data.FieldSpec;
 
-import static org.apache.pinot.controller.recommender.rules.io.params.RecommenderConstants.DEFAULT_AVERAGE_NUM_VALUES_PER_ENTRY;
-import static org.apache.pinot.controller.recommender.rules.io.params.RecommenderConstants.DEFAULT_CARDINALITY;
-import static org.apache.pinot.controller.recommender.rules.io.params.RecommenderConstants.DEFAULT_DATA_LENGTH;
+import static org.apache.pinot.common.data.schema.SchemaMetadataConstants.DEFAULT_AVERAGE_NUM_VALUES_PER_ENTRY;
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/data/schema/FieldMetadata.java:26]: It is preferable to have all the recommender code in a sub-directory as it is currently organized in pinot/controller/recommender/...

So, please consider undoing this directory move

File: pinot-common/src/main/java/org/apache/pinot/common/data/schema/SchemaMetadataConstants.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.common.data.schema;
+
+public class SchemaMetadataConstants {
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/data/schema/SchemaMetadataConstants.java:22]: This should be in RecommenderConstants

File: pinot-common/src/main/java/org/apache/pinot/common/data/schema/DateTimeFieldSpecMetadata.java [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.common.data.schema;
+
+
+public class DateTimeFieldSpecMetadata extends FieldMetadata {
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/data/schema/DateTimeFieldSpecMetadata.java:23]: Discussed offline with @sajjad-moradi  on why this is needed. Please add some javadoc explaining the context on why this is needed.

**Review by siddharthteotia — APPROVED** [review]
LGTM. Thanks for this change
Just left some minor comments

File: pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/NumberGenerator.java (hunk 13 above)
Comment (siddharthteotia) [reply pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/NumberGenerator.java]: Got it. Thanks

File: pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java (hunk 11 above)
Comment (siddharthteotia) [reply pinot-tools/src/main/java/org/apache/pinot/tools/realtime/provisioning/MemoryEstimator.java]: Just wanted to ensure that the new code (added in this PR) to generated the completed segment generates it in the correct directory hierarchy as expected by the second constructor. Looks like your unit test already covers it.

File: pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/StringGenerator.java (hunk 12 above)
Comment (sajjad-moradi) [reply pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/StringGenerator.java:75]: That's correct, but even with completed segments, `MemoryEstimator` tries to extrapolate the actual memory size based on secondsToConsume (derived from numHours input) parameter:
```java
int totalDocs = (int) (((double) secondsToConsume / _sampleSegmentConsumedSeconds) * _totalDocsInSampleSegment);
long memoryForConsumingSegmentPerPartition = getMemoryForConsumingSegmentPerPartition(statsFile, totalDocs);
```
My point we have only one completed segment (given or generated) and that should be enough to for estimating memory for different consumption window sizes.
